### ISSUE TITLE
JWT in localStorage

### ISSUE THREAD

**Issue by sedubois** [strapi/strapi-sdk-javascript#31]
Issue #7 mentions that JWTs are stored in localStorage, but according to [this article](https://www.rdegges.com/2018/please-stop-using-local-storage/), this is subject to XSS attacks so instead they should be stored in cookies.

> please, please, whatever you do, do not store session information (like JSON Web Tokens) in local storage. This is a very bad idea and will open you up to an extremely wide array of attacks that could absolutely cripple your users.

Do you agree with this? What should be done? I also think it's important to educate everyone to secure practices. Thanks for building Strapi and keep it up!

- [x] I'm sure that this issue hasn't already been referenced

**Comment by lucaperret** [comment]
Hello @sedubois,
You can disable the localStorage store by setting it to false:
```js
const strapi = new Strapi('http://localhost:1337', {
  localStorage: false
});
```

**Comment by lucaperret** [comment]
Hi @sedubois, is it solving your problem ?

**Comment by pierreburgy** [comment]
According to many articles, storing tokens in local storage is an important security mistake:
 - https://auth0.com/docs/security/store-tokens
 - https://stormpath.com/blog/where-to-store-your-jwts-cookies-vs-html5-web-storage
 - https://dev.to/rdegges/please-stop-using-local-storage-1i04

But, tell me if I am wrong, the `js-cookie` module causes the exact same issue, because the token is accessible through any script running on the same domain. The only way to solve this issue is to store the token from a server with `HttpOnly`: https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#Secure_and_HttpOnly_cookies.

**Comment by sedubois** [comment]
Yes as far as I know, the secure practice is to use httpOnly cookies. JWTs should only be considered for non-browser apps (which are not subject to XSS).

**Comment by pmarrone** [comment]
Hi there. I was trying to figure out how to work around this. Please, let me know if this makes sense.

In the case that we wanted to store the jwt in an httpOnly cookie, the immediate implication would be that there would be no way to add the Authentication header via JS. Instead, the cookie would be send with each request to the API. This means that the API would need a way to extract the token from the cookie as an alternative to the authentication header.

The users-permissions module doesn't seem to be prepared for this out of the box. For the httpOnly cookie to work, I guess that this could work by either
1. Modify the _permissions_ policy and the _JWT_ service to also check for some cookie (configurable?) for the authentication token (say, a cookie named jwt).
2. Adding middleware to modify the original request's headers so that the jwt cookie is set as the authentication header (which seems kind of a hack).

In any case, if an attacker manages to inject JS code, what would prevent him/her from making calls in the user's behalf, even if he can't see the contents of the token? If the token doesn't hold sensitive information in itself, I fail to see how stealing the token would be more dangerous than making requests directly from the hijacked browser.

**Comment by sedubois** [comment]
Thanks @pmarrone, I reported my observation based on my limited understanding of the issue. Hopefully a security expert (such as the authors of the above linked articles) could provide recommendations.
